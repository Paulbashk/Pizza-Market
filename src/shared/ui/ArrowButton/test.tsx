import { screen } from '@testing-library/react';
import { renderApp } from '@/test/utils';
import { ArrowButton } from '.';

describe('UI ArrowButton', () => {
  it('should render', () => {
    const { container } = renderApp(<ArrowButton />, {});

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toBeInTheDocument();
  });

  it('should render with props example className', () => {
    const { container } = renderApp(<ArrowButton className="test-class" />, {});

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toHaveClass('test-class');
  });
});
