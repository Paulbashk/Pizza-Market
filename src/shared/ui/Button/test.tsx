import { screen } from '@testing-library/react';
import { theme } from '@/shared/libs/assets/styles';
import { renderApp } from '@/test/utils';
import { Button, ButtonColors } from '.';

describe('UI Button', () => {
  it('should render with content', () => {
    const { container } = renderApp(
      <Button color={ButtonColors.gray} isActive={false}>
        Привет мир
      </Button>,
      {},
    );

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('Привет мир');
  });

  it('should render gray hovered inactive', () => {
    const { container } = renderApp(
      <Button color={ButtonColors.gray} isActive={false}>
        Привет мир
      </Button>,
      {},
    );

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toHaveStyle(`background-color: ${theme.colors.lightgray}`);
    expect(button).toHaveStyle(`color: ${theme.colors.dark}`);

    expect(button).toHaveStyleRule('background-color', theme.colors.dark, {
      modifier: ':hover',
    });
    expect(button).toHaveStyleRule('color', theme.colors.white, {
      modifier: ':hover',
    });
  });

  it('should render gray hovered active', () => {
    const { container } = renderApp(
      <Button color={ButtonColors.gray} isActive>
        Привет мир
      </Button>,
      {},
    );

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toHaveStyle(`background-color:${theme.colors.dark}`);
    expect(button).toHaveStyle(`color: ${theme.colors.white}`);
    expect(button).toHaveStyle('opacity: 1');

    expect(button).toHaveStyleRule('opacity', '0.8', {
      modifier: ':hover',
    });
  });

  it('should render transparent hovered inactive', () => {
    const { container } = renderApp(
      <Button color={ButtonColors.transparent} isActive={false}>
        Привет мир
      </Button>,
      {},
    );

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toHaveStyle('background-color: transparent');
    expect(button).toHaveStyle(`color: ${theme.colors.yellow}`);

    expect(button).toHaveStyleRule('background-color', theme.colors.yellow, {
      modifier: ':hover',
    });
    expect(button).toHaveStyleRule('color', theme.colors.white, {
      modifier: ':hover',
    });
  });

  it('should render transparent hovered active', () => {
    const { container } = renderApp(
      <Button color={ButtonColors.transparent} isActive>
        Привет мир
      </Button>,
      {},
    );

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toHaveStyle('background-color: transparent');
    expect(button).toHaveStyle(`color: ${theme.colors.yellow}`);

    expect(button).toHaveStyleRule('background-color', theme.colors.yellow, {
      modifier: ':hover',
    });
    expect(button).toHaveStyleRule('color', theme.colors.white, {
      modifier: ':hover',
    });
  });

  it('should render with props example className', () => {
    const { container } = renderApp(
      <Button color={ButtonColors.transparent} isActive className="test-class">
        Привет мир
      </Button>,
      {},
    );

    expect(container).toMatchSnapshot();

    const button = screen.getByRole('button');

    expect(button).toHaveClass('test-class');
  });

  it('should accept invalid child', () => {
    renderApp(
      <Button color={ButtonColors.gray} isActive>
        Привет
        {null}
      </Button>,
      {},
    );
  });
});
