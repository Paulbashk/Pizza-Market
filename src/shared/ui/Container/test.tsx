import { renderApp } from '@/test/utils';
import Container from './index';

describe('UI Container', () => {
  it('should render with content', () => {
    const { container } = renderApp(<Container>Content</Container>, {});

    expect(container).toMatchSnapshot();
    expect(container.firstChild).toBeInTheDocument();
  });

  it('should render with props example className', () => {
    const { container } = renderApp(
      <Container className="container-test">Content</Container>,
      {},
    );

    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveClass('container-test');
  });

  it('should accept invalid child', () => {
    renderApp(
      <Container>
        <p /> {null}
      </Container>,
      {},
    );
  });
});
