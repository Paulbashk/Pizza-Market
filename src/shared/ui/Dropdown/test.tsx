import { renderApp } from '@/test/utils';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Dropdown from './index';

jest.mock('hex-rgb', () => () => '');

const dropdownItems = [
  {
    value: 'item1',
    label: 'Элемент 1',
  },
  {
    value: 'item2',
    label: 'Элемент 2',
  },
  {
    value: 'item3',
    label: 'Элемент 3',
  },
];

const DropdownLabel = ({ label, role }: { label: string; role?: string }) => (
  <span role={role}>{label}</span>
);

describe('UI Dropdown', () => {
  it('should display with default props and menu on hover', async () => {
    const user = userEvent.setup();

    const { container } = renderApp(
      <Dropdown
        trigger={<DropdownLabel label={dropdownItems[0].label} />}
        items={dropdownItems.map(item => (
          <span key={item.value}>{item.label}</span>
        ))}
        selected={0}
      />,
      {},
    );

    expect(container).toMatchSnapshot();

    const label = screen.getByRole('label');

    expect(label).toBeInTheDocument();
    expect(label).toHaveTextContent(dropdownItems[0].label);

    expect(screen.queryByRole('listbox')).toBeNull();

    await user.hover(label);

    expect(screen.queryByRole('listbox')).toBeInTheDocument();
    expect(screen.queryAllByRole('option').length).toBe(3);
  });

  it('should update label and state when option is selected', async () => {
    const user = userEvent.setup();

    const { setProp } = renderApp(
      <Dropdown
        trigger={<DropdownLabel label={dropdownItems[0].label} />}
        items={dropdownItems.map(item => (
          <span key={item.value}>{item.label}</span>
        ))}
        selected={0}
      />,
      {},
    );

    await user.hover(screen.getByRole('label'));
    await user.click(screen.getAllByRole('option')[1]);

    setProp({
      trigger: <DropdownLabel label={dropdownItems[1].label} />,
      items: dropdownItems.map(item => (
        <span key={item.value}>{item.label}</span>
      )),
      selected: 1,
    });

    expect(screen.queryByRole('listbox')).toBeNull();
    expect(screen.getByRole('label')).toHaveTextContent(dropdownItems[1].label);
  });

  it('checking prop: without selected', async () => {
    const user = userEvent.setup();

    renderApp(
      <Dropdown
        trigger={<DropdownLabel label={dropdownItems[0].label} />}
        items={dropdownItems.map(item => (
          <span key={item.value}>{item.label}</span>
        ))}
      />,
      {},
    );

    const label = screen.getByRole('label');

    expect(label).toBeInTheDocument();
    expect(label).toHaveTextContent(dropdownItems[0].label);

    expect(screen.queryByRole('listbox')).toBeNull();

    await user.hover(label);

    expect(screen.queryByRole('listbox')).toBeInTheDocument();
    expect(screen.queryAllByRole('option').length).toBe(3);
  });

  it('checking prop: onOpen', async () => {
    const user = userEvent.setup();

    const onMockOpen = jest.fn((open: boolean) => !open);

    renderApp(
      <Dropdown
        trigger={<DropdownLabel label={dropdownItems[0].label} />}
        items={dropdownItems.map(item => (
          <span key={item.value}>{item.label}</span>
        ))}
        selected={0}
        onOpen={onMockOpen}
      />,
      {},
    );

    await user.hover(screen.getByRole('label'));

    expect(onMockOpen.mock.calls).toHaveLength(1);
    expect(onMockOpen.mock.results[0].value).toBe(false);

    expect(screen.queryByRole('listbox')).toBeNull();
  });

  it('checking prop: onSelected', async () => {
    const user = userEvent.setup();

    const onMockSelected = jest.fn((id: number) => id);

    renderApp(
      <Dropdown
        trigger={<DropdownLabel label={dropdownItems[0].label} />}
        items={dropdownItems.map(item => (
          <span key={item.value}>{item.label}</span>
        ))}
        selected={0}
        onSelected={onMockSelected}
      />,
      {},
    );

    expect(onMockSelected.mock.calls).toHaveLength(0);

    await user.hover(screen.getByRole('label'));
    await user.click(screen.getAllByRole('option')[1]);

    expect(onMockSelected.mock.calls).toHaveLength(1);
    expect(onMockSelected.mock.results[0].value).toBe(1);
  });

  it('check prop: items are empty', async () => {
    const user = userEvent.setup();

    const { container } = renderApp(
      <Dropdown
        trigger={<DropdownLabel label={dropdownItems[0].label} />}
        items={[]}
        selected={0}
      />,
      {},
    );

    expect(container).toMatchSnapshot();

    await user.hover(screen.getByRole('label'));
    expect(screen.queryByRole('listbox')).toBeNull();
  });
});
