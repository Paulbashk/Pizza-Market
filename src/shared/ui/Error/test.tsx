import { renderApp } from '@/test/utils';
import { screen } from '@testing-library/react';
import Error from './index';

describe('UI Error', () => {
  it('should render', () => {
    const { container } = renderApp(<Error>Error page</Error>, {});

    expect(container).toMatchSnapshot();

    expect(screen.getByText(/Error page/)).toBeInTheDocument();
  });

  it('should render with props example className', () => {
    const { container } = renderApp(
      <Error className="error-test">Error page</Error>,
      {},
    );

    expect(container).toMatchSnapshot();

    const error = screen.getByText(/Error page/);

    expect(screen.getByText(/Error page/)).toBeInTheDocument();
    expect(error).toHaveClass('error-test');
  });

  it('should accept invalid child', () => {
    renderApp(
      <Error>
        Text
        {null}
      </Error>,
      {},
    );
  });
});
