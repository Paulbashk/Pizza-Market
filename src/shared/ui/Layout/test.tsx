import React from 'react';
import { renderApp } from '@/test/utils';
import { screen } from '@testing-library/react';
import Layout from './index';

jest.mock('next/head', () => ({
  __esModule: true,
  default: ({ children }: { children: Array<React.ReactElement> }) => children,
}));

describe('UI Layout', () => {
  it('should render', () => {
    const title = 'title page';

    const { container } = renderApp(<Layout title={title}>Content</Layout>, {});

    expect(container).toMatchSnapshot();

    expect(screen.getByText(title)).toBeInTheDocument();

    expect(screen.getByText(/Content/)).toBeInTheDocument();
  });
});
