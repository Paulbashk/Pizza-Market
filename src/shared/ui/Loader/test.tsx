import { renderApp } from '@/test/utils';
import { screen } from '@testing-library/react';
import { Loader } from '.';

describe('UI Loader', () => {
  it('should render nothing is loading prop is false', () => {
    const { container } = renderApp(<Loader loading={false} />, {});

    expect(container).toMatchSnapshot();
    expect(container.firstChild).toBeNull();
  });

  it('should have allow style override on wrapper', () => {
    const style = { overflow: 'scroll' };
    const { container } = renderApp(<Loader cssOverride={style} />, {});

    expect(container).toMatchSnapshot();
    expect(container.firstChild).toHaveStyle(style);
  });

  it('should have allow custom html props', () => {
    const { container } = renderApp(<Loader aria-label="aria-label" />, {});

    expect(container).toMatchSnapshot();
    expect(screen.queryByLabelText('aria-label')).toBeTruthy();
  });
});
