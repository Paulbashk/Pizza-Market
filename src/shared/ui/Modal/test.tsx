/* eslint-disable max-len */
import { renderApp } from '@/test/utils';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Modal, ModalWindow } from '.';

describe('UI Modal', () => {
  it('should dont display anything, isOpen is false', () => {
    const { container } = renderApp(
      <Modal isOpen={false}>
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    expect(container.firstChild).toBeNull();
  });

  it('should render and display child content, isOpen is true', () => {
    renderApp(
      <Modal isOpen>
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    expect(screen.getByText('Modal Content')).toBeInTheDocument();
  });

  it('prop check: allowScroll is true, there will be no overflow hidden in the tag body', async () => {
    const style = 'overflow: hidden';

    const { setProp } = renderApp(
      <Modal isOpen allowScroll>
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    expect(document.body).not.toHaveStyle(style);

    setProp({
      isOpen: false,
      allowScroll: true,
    });

    expect(global.document.body).not.toHaveStyle(style);
  });

  it('prop check: allowScroll is false, the body tag will have overflow hidden when the window is shown', async () => {
    const style = 'overflow: hidden';

    const { setProp } = renderApp(
      <Modal isOpen allowScroll={false}>
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    expect(document.body).toHaveStyle(style);

    setProp({
      isOpen: false,
      allowScroll: false,
    });

    expect(global.document.body).not.toHaveStyle(style);
  });

  it('prop check: when you press the Escape key, the onEscapeKeydown function will work', async () => {
    const user = userEvent.setup();
    const onMockEscape = jest.fn();

    renderApp(
      <Modal isOpen onEscapeKeydown={onMockEscape}>
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    await user.keyboard('{Escape}');

    expect(onMockEscape.mock.calls).toHaveLength(1);
  });

  it('prop check: beforeOpen and afterOpen', () => {
    const onMockBeforeOpen = jest.fn();
    const onMockAfterOpen = jest.fn();

    const { setProp } = renderApp(
      <Modal isOpen afterOpen={onMockAfterOpen} beforeOpen={onMockBeforeOpen}>
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    expect(onMockAfterOpen.mock.calls).toHaveLength(1);
    expect(onMockBeforeOpen.mock.calls).toHaveLength(0);

    setProp({ isOpen: false });

    expect(onMockAfterOpen.mock.calls).toHaveLength(1);
    expect(onMockBeforeOpen.mock.calls).toHaveLength(1);
  });

  it('prop check: beforeClose and afterClose', () => {
    const onMockBeforeClose = jest.fn();
    const onMockAfterClose = jest.fn();

    const { setProp } = renderApp(
      <Modal
        isOpen
        afterClose={onMockAfterClose}
        beforeClose={onMockBeforeClose}
      >
        <ModalWindow>Modal Content</ModalWindow>
      </Modal>,
      {},
    );

    expect(onMockAfterClose.mock.calls).toHaveLength(0);
    expect(onMockBeforeClose.mock.calls).toHaveLength(1);

    setProp({ isOpen: false });

    expect(onMockAfterClose.mock.calls).toHaveLength(1);
    expect(onMockBeforeClose.mock.calls).toHaveLength(1);
  });

  it('should accept invalid child', () => {
    renderApp(<Modal isOpen>{null}</Modal>, {});
  });
});
