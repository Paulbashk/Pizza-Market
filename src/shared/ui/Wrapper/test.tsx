import { renderApp } from '@/test/utils';
import { screen } from '@testing-library/react';
import React from 'react';
import { Wrapper } from '.';

jest.mock('hex-rgb', () => () => '');

jest.mock('swiper/scss', jest.fn());
jest.mock('swiper/scss/effect-coverflow', jest.fn());
jest.mock('swiper/scss/pagination', jest.fn());
jest.mock('swiper/scss/navigation', jest.fn());
jest.mock('swiper/scss/autoplay', jest.fn());

jest.mock('swiper/react', () => ({
  Swiper: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="Swiper-testId">{children}</div>
  ),
  SwiperSlide: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="SwiperSlide-testId">{children}</div>
  ),
}));

jest.mock('swiper', () => ({
  Navigation: () => null,
  Pagination: () => null,
  Scrollbar: () => null,
  A11y: () => null,
}));

describe('Wrapper', () => {
  it('should render with children', () => {
    const { container } = renderApp(
      <Wrapper>
        <p>Content</p>
      </Wrapper>,
      {},
    );

    expect(container).toMatchSnapshot();

    const content = screen.getByText('Content');
    expect(content).toBeInTheDocument();
  });

  it('should render with other prop', () => {
    const idWrapper = 'test-id';

    const { container } = renderApp(
      <Wrapper id={idWrapper} data-testid="wrapper">
        Text
      </Wrapper>,
    );

    expect(container).toMatchSnapshot();

    expect(screen.getByTestId('wrapper')).toHaveAttribute('id', idWrapper);
  });

  it('should accept invalid child', () => {
    renderApp(
      <Wrapper>
        <p />
        {null}
      </Wrapper>,
    );
  });
});
