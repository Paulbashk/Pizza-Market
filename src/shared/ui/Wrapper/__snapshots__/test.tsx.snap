// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper should render with children 1`] = `
.c0 {
  background-color: #fff;
  border-radius: 16px;
  box-shadow: 0px 0px 15px rgba(0,0,0,0.25);
}

<div>
  <div
    class="c0"
  >
    <p>
      Content
    </p>
  </div>
</div>
`;

exports[`Wrapper should render with other prop 1`] = `
.c0 {
  background-color: #fff;
  border-radius: 16px;
  box-shadow: 0px 0px 15px rgba(0,0,0,0.25);
}

<div>
  <div
    class="c0"
    data-testid="wrapper"
    id="test-id"
  >
    Text
  </div>
</div>
`;
